src/daft-logical-plan/src/optimization/rules/push_down_anti_semi_join.rs

```diff
@@ -0,0 +1,673 @@
+///
+/// Anti and semi joins work similarly to filters, where the left side is the input and the right side is the filter predicate.
+/// They will only eliminate columns from the left side so, pushing them down will reduce the intermediate cardinalities of our data,
+/// generally improving query performance.
+///
+/// One case where it may not improve query performance is when the join is pushed down below a filter.
+/// However, we should still do it here so that we can potentially push it through joins below the filter, and then rely on PushDownFilter
+/// to then move filters below this join again.
+///
+/// In the future, we should incorporate anti and semi join reordering into our general join reordering algorithm.
+///
+/// # Examples
+///
+/// ## Through a project
+///
+/// Before:
+/// ```text
+/// * Semi Join
+/// |   on: x == a
+/// |\
+/// | * Scan
+/// | |   schema: (a,)
+/// |
+/// * Project
+/// |   projection: y AS x
+/// |
+/// * Scan
+/// |   schema: (y,)
+/// ```
+///
+/// After:
+/// ```text
+/// * Project
+/// |   projection: y AS x
+/// |
+/// * Semi Join
+/// |   on: y == a
+/// |\
+/// | * Scan
+/// | |   schema: (a,)
+/// |
+/// * Scan
+/// |   schema: (y,)
+/// ```
+///
+/// ## Through an inner join
+///
+/// Before:
+/// ```text
+/// * Anti Join
+/// |   on: x == a
+/// |\
+/// | * Scan
+/// | |   schema: (a,)
+/// |
+/// * Inner Join
+/// |   on: x == y
+/// |\
+/// | * Scan
+/// | |   schema: (y,)
+/// |
+/// * Scan
+/// |   schema: (x,)
+/// ```
+///
+/// After:
+/// ```text
+/// * Inner Join
+/// |   on: x == y
+/// |\
+/// | * Scan
+/// | |   schema: (y,)
+/// |
+/// * Anti Join
+/// |   on: x == a
+/// |\
+/// | * Scan
+/// | |   schema: (a,)
+/// |
+/// * Scan
+/// |   schema: (x,)
+/// ```
+#[derive(Debug)]
+pub struct PushDownAntiSemiJoin {}
+
+impl PushDownAntiSemiJoin {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for PushDownAntiSemiJoin {
+    fn try_optimize(&self, plan: LogicalPlanRef) -> DaftResult<Transformed<LogicalPlanRef>> {
+        plan.transform_down(|node| {
+            if let LogicalPlan::Join(Join {
+                left,
+                right,
+                on,
+                join_type,
+                join_strategy,
+                ..
+            }) = node.as_ref()
+                && matches!(join_type, JoinType::Anti | JoinType::Semi)
+            {
+                match left.as_ref() {
+                    // We can push an anti/semi join down a project if it the columns the join uses (or its aliases) are present in the input of the project.
+                    LogicalPlan::Project(Project {
+                        input, projection, ..
+                    }) => {
+                        let Some(on_expr) = on.inner() else {
+                            return Ok(Transformed::no(node));
+                        };
+
+                        let trivial_projections_map = projection
```

Comment:
this seems like fairly common code. It may be helpful to put the projection expr list into a  `Projection` type and add a `to_projection_map` or something

Reply:
Yeah that makes sense. This will definitely need to be reworked in our ordinal refactor and I don't plan on making new optimization rules before I get some of that through, so I think I will keep this for now, but I will make sure to pull this and other similar logic out into functions and use them in our other rules as well.